File: src/python-app/titanic_analysis.py
'''
14.Explore, add, and adjust the data as you see fit. 
15. Build a logistic regression model to predict survivability on the training set using any features 
that you see fit.
16. Measure the accuracy of your model on the training set.
17. Load `test.csv` and predict your model on the test set.
18. Measure the accuracy of your model on the test set.
19. For questions 14-18, ensure you have print statements on all changes that you make. 
This will show through the terminal all of your different changes. 
The grader does not need to see your code in your `.py` script - only the outputs.'''


import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

df = pd.read_csv("src/data/train.csv")
print('View the head of the dataset: ', df.head())
print('Dataset description:', df.describe())
print('Dataset columns: ', df.columns)

# Dealing the missing value and nan value
print("View the missing value: ", df.isna().sum())
print('Imputation the unknown age with the mean value.')
df['Age'] = df['Age'].fillna(df['Age'].mean())
print('Since there are too many unknown value for cabin, I will not include this column in the further analysis.')
df = df.drop('Cabin',axis = 1)
df = df.dropna(subset =['Embarked'])
print('Drop 2 rows that the Embarked value is nan.')

print('Encoding the categorical variables (sex and embarked) to numeric value.')
df = pd.get_dummies(df, columns = ['Sex', 'Embarked'], drop_first=True)

# select the features to fit the model 
print(df.corr(numeric_only=True))
print("Since the correlation shows that the survivability related to the Pclass, age, SibSp, parch, fare, sex, and embarked, I will use these variables to fit the logistic model.")

features = ['Pclass', 'Age','SibSp','Parch','Fare','Sex_male','Embarked_Q','Embarked_S']
x_train = df[features]
y_train = df['Survived']
model = LogisticRegression()
model.fit(x_train,y_train)
train_acc = model.score(x_train,y_train)
print(f"Training accuracy of the model is {train_acc:.3f}")

df_test = pd.read_csv("src/data/test.csv")
print("View the missing value: ", df_test.isna().sum())
print('Imputation the unknown age with the mean value.')
df_test['Age'] = df_test['Age'].fillna(df_test['Age'].mean())
df_test = df_test.dropna(subset =['Fare'])
print('Drop 1 row that the Fare value is nan.')
print('Encoding the categorical variables (sex and embarked) to numeric value.')
df_test = pd.get_dummies(df_test, columns = ['Sex', 'Embarked'], drop_first=True)

x_test = df_test[features]
y_pred = model.predict(x_test)
print("The predicted value on the test data: ", y_pred)

output = pd.DataFrame({
    "y_pred": y_pred
})

output.to_csv("src/data/predictions_python.csv", index=False)
print("The prediction was saved successfully into the predictions.csv file.")

File: src/R-app/titanic_analysis.R
# Load libraries
library(readr)
library(dplyr)

# prepocess the data
df <- read_csv("src/data/train.csv")
cat("View the head of the dataset:\n")
print(head(df))
cat("\nDataset description:\n")
print(summary(df))
cat("\nDataset columns:\n")
print(colnames(df))

# deal with the missing value
cat("\nView the missing value count:\n")
print(colSums(is.na(df)))

cat("\nImpute unknown Age with the mean value.\n")
df$Age[is.na(df$Age)] <- mean(df$Age, na.rm = TRUE)

cat("\nSince Cabin has too many missing values, drop it.\n")
df <- df %>% select(-Cabin)

cat("\nDrop rows where Embarked is missing.\n")
df <- df %>% filter(!is.na(Embarked))

# encoding the categorical variables
cat("\nEncoding categorical variables (Sex and Embarked) to numeric dummies.\n")

# Convert to factors
df$Sex <- as.factor(df$Sex)
df$Embarked <- as.factor(df$Embarked)

# One-hot encode automatically in model.matrix
df_encoded <- model.matrix(~ Pclass + Age + SibSp + Parch + Fare + Sex + Embarked, data = df)[, -1]

cat("\nCorrelation matrix for numeric variables:\n")
print(cor(df_encoded))

cat("\nSince correlation shows that survivability relates to several variables, we'll use them for logistic regression.\n")

# fit logistic regression
features <- c("Pclass", "Age", "SibSp", "Parch", "Fare", "Sexmale", "EmbarkedQ", "EmbarkedS")

x_train <- df_encoded[, features]
y_train <- df$Survived

# Build logistic regression model
cat("\nFitting logistic regression model on training data...\n")
train_data <- as.data.frame(cbind(y_train, x_train))
model <- glm(y_train ~ ., data = train_data, family = binomial())

# Training accuracy
train_prob <- predict(model, type = "response")
train_pred <- ifelse(train_prob > 0.5, 1, 0)
train_acc <- mean(train_pred == y_train)
cat(sprintf("Training accuracy of the model is %.3f\n", train_acc))

# load the test dataset and predict
df_test <- read_csv("src/data/test.csv")
cat("\nView the missing value count in test data:\n")
print(colSums(is.na(df_test)))

cat("\nImpute unknown Age with mean value.\n")
df_test$Age[is.na(df_test$Age)] <- mean(df_test$Age, na.rm = TRUE)

cat("\nDrop rows with missing Fare.\n")
df_test <- df_test %>% filter(!is.na(Fare))

cat("\nEncoding categorical variables (Sex and Embarked) to numeric dummies in test data.\n")
df_test$Sex <- as.factor(df_test$Sex)
df_test$Embarked <- as.factor(df_test$Embarked)

# Use same dummy variables as training
df_test_encoded <- model.matrix(~ Pclass + Age + SibSp + Parch + Fare + Sex + Embarked, data = df_test)[, -1]

# Align columns (ensure same feature order)
common_cols <- intersect(colnames(df_encoded), colnames(df_test_encoded))
df_test_encoded <- df_test_encoded[, common_cols, drop = FALSE]

cat("\nPredicting survivability on test set...\n")
y_pred <- predict(model, newdata = as.data.frame(df_test_encoded), type = "response")
y_pred_class <- ifelse(y_pred > 0.5, 1, 0)
cat("Predicted values on the test data:\n")
print(head(y_pred_class))

output <- data.frame(y_pred = y_pred)
# Save the data frame to a CSV file (no row names)
write.csv(output, "src/data/predictions_R.csv", row.names = FALSE)
# Print confirmation message
cat("The prediction was saved successfully into the predictions.csv file.\n")
